lib/dpl/provider/pages.rb

```diff
@@ -44,34 +53,92 @@ def slug
         options.fetch(:repo) { context.env['TRAVIS_REPO_SLUG'] }
       end
 
+      def keep_history
+        options.fetch(:keep_history, false)
+      end
+
+      def allow_empty_commit
+        options.fetch(:allow_empty_commit, false)
+      end
+
+      def verbose
+        # Achtung! Never verbosify git, since it may expose user's token.
+        options.fetch(:verbose, false)
+      end
+
       def check_auth
       end
 
       def needs_key?
         false
       end
 
-      def github_deploy
-        context.shell 'rm -rf .git > /dev/null 2>&1'
-        context.shell "touch \"deployed at `date` by #{@gh_name}\""
-        context.shell 'git init' or raise 'Could not create new git repo'
+      def print_step(msg)
+        log msg if @verbose
+      end
+
+      def github_pull_or_init(target_dir)
+        unless @keep_history
+          github_init
+          return
+        end
+
+        print_step "Trying to clone a single branch #{@target_branch} from existing repo..."
+        unless context.shell "git clone --quiet --branch='#{@target_branch}' --depth=1 '#{@gh_remote_url}' '#{target_dir}' &>/dev/null"
+          # if such branch doesn't exist at remote, init it from scratch
+          print_step "Cloning #{@target_branch} branch failed"
+          Dir.mkdir(target_dir)  # Restore dir destroyed by failed `git clone`
```

Comment:
Because this is the directory created by mktmpdir, it's inside a world-writable directory. A concurrent use of mktmpdir could have taken ownership of the directory between when the clone deleted it and when this line tries to recreate it. That's why my patch makes this a subdirectory of the one created by mktmpdir.

Reply:
done